---
title: "Oct. 23"
author: "Mio Shibata"
date: "2018/10/29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Make 1x3 table of SNP three genotypes. Calculate sample allele frequency.

```{r}
a.freq <- 0.4
a.freqs <- c(a.freq, 1-a.freq) # Frequency of two alleles. 
g.freqs.hwe <- c(a.freqs[1]^2, 2*a.freqs[1]*a.freqs[2], a.freqs[2]^2)

N <- 100
g.obs.data <- sample(1:3,N,replace=TRUE,prob=g.freqs.hwe)
g.obs <- tabulate(g.obs.data,3)
g.obs
```


```{r}
a.obs <- (g.obs[1]*2 + g.obs[2])/(sum(g.obs)*2)
a.obss <- c(a.obs,1-a.obs)
a.obss[1]
```
Generate a series of 1x3 tables that keeps the observed allele counts.
```{r}
my_3genotype <- function (X){
  tmp.X <- X
  n <- sum(tmp.X)
  if (tmp.X[1] > tmp.X[3]){
    tmp.X[1] <- X[3]
    tmp.X[3] <- X[1]
    }
  n_0 <- 2*tmp.X[1]+tmp.X[2] # minor allele
  n_1 <- 2*tmp.X[3]+tmp.X[2]
  range_0 <- 0:(floor(n_0/2))
  ret <- matrix(0,length(range_0),3)
  ret[,1] <- range_0
  ret [ ,2] <- n_0-2*range_0
  ret[,3] <- n-(ret[,1]+ret[,2])
  ret
}
series_table <- my_3genotype( g.obs )
```

I generated the table which is stored each individual's allels and make a series of tables which are produced by the 1x3 tables.

```{r}
my.prob.hwe <- function(X){
  g.freq <- c((X[1]/sum(X)),(X[2]/sum(X)),(X[3]/sum(X)))
  log.prob <- lgamma(sum(X)+1)-sum(lgamma(X+1)) + sum(X * log(g.freq))
  prob <- exp(log.prob)
  return(prob)
}

table.prob <- function(t){
 marg.row <- apply(t,1,sum)
 marg.col <- apply(t,2,sum)
 N <- sum(t)
 prob.log <- sum(lgamma(marg.row+1)) + sum(lgamma(marg.col+1)) -(lgamma(N+1)+sum(lgamma(t+1)))
 prob <- exp(prob.log)
 return(prob)
}

hwe.table <- function(X){
 allele.pattern.homo.1 <- c(2,0)
 allele.pattern.hetero <- c(1,1)
 allele.pattern.homo.2 <- c(0,2)
 m <- matrix(c(rep(allele.pattern.homo.1, X[1]),rep(allele.pattern.hetero, X[2]),rep(allele.pattern.homo.2,X[3])),byrow =TRUE, ncol = 2)
 return(m)
}

prob.hwe <- function(X){
 hwe.tbl <- hwe.table(X)
 prob <- table.prob(hwe.tbl)
 prob2.log <- lgamma(sum(X)+1)-sum(lgamma(X+1))
 prob <- prob * exp(prob2.log)
 return(prob)
}


##Generating a series of table and store the p.value by HWE test.
pr <- rep(0,length(series_table[,1]))
ori.pr <- rep(0,length(series_table[,1]))
for(i in 1:length(pr)){
    pr[i] <-prob.hwe(series_table[i,])
    ori.pr[i] <- my.prob.hwe(series_table[i,])
}
result <- list(probs = pr,p.value=sum(pr[which(pr<=ori.pr)]))

result
sum(result$probs)
```

